### PR TITLE
Reified optics for Projection Expressions

### PR DESCRIPTION
**Pull request by googley42**
This PR introduces reified optics for projection expressions. This enables type safe queries and updates and deletes, plugging the type safety gaps that were previously there.

### CODE REVIEW

File: dynamodb/src/main/scala/zio/dynamodb/ProjectionExpression.scala [hunk 1]
```diff
@@ -65,15 +74,22 @@ sealed trait ProjectionExpression { self =>
...
-  def contains[A](av: A)(implicit t: ToAttributeValue[A]): ConditionExpression                   =
+  // constraint: String OR Set
+  def contains[A](av: A)(implicit t: ToAttributeValue[A]): ConditionExpression =
     ConditionExpression.Contains(self, t.toAttributeValue(av))
-  def beginsWith[A](av: A)(implicit t: ToAttributeValue[A]): ConditionExpression                 =
-    ConditionExpression.BeginsWith(self, t.toAttributeValue(av))
+
+  // constraint: String
+  def beginsWith(av: String)(implicit ev: RefersToString[To]): ConditionExpression = {
+    val _ = ev
```
Comment (nhyne) [inline dynamodb/src/main/scala/zio/dynamodb/ProjectionExpression.scala:88]: Is this just an ignored value being set? Do we need this here?

**Review by nhyne — APPROVED** [review]
This is awesome work! Great Job! 

I'm not super familiar with the Optics work so I may not be the best person for approval for that but the examples + tests make sense to me.

File: dynamodb/src/main/scala/zio/dynamodb/ProjectionExpression.scala (hunk 1 above)
Comment (googley42) [reply dynamodb/src/main/scala/zio/dynamodb/ProjectionExpression.scala:88]: this is a compile time proof to force type safety, we don't actually use `ev`

Comment (nhyne) [reply dynamodb/src/main/scala/zio/dynamodb/ProjectionExpression.scala:88]: Ah okay that's really cool! I don't I've seen one of these before.

**Comment by googley42** [comment]
closes https://github.com/zio/zio-dynamodb/issues/86

**Pull request merged**